src/covid_model_seiir_pipeline/io/keys.py

```diff
@@ -0,0 +1,124 @@
+
+    """
+    DRAW_TEMPLATE: str = 'draw_{draw_id}'
+    MEASURE_TEMPLATE: str = '{measure}'
+    COV_SCENARIO_TEMPLATE: str = '{covariate_scenario}_scenario'
+    COV_INFO_TEMPLATE: str = '{info_type}_info'
+
+
+LEAF_TEMPLATES = __LeafTemplates()
+
+
+class __PathTemplates(NamedTuple):
+    SCENARIO_TEMPLATE: str = '{scenario}'
+
+
+PATH_TEMPLATES = __PathTemplates()
+
+
+class DatasetKey(NamedTuple):
+    """Struct representing metadata for an on-disk dataset.
+
+    Attributes
+    ==========
+    root
+        The path to the parent directory in which this dataset lives or will
+        live.
+    output_format
+        The format of the dataset on disk.
+    data_type
+        The type of data being stored. This should be a logical type like
+        `deaths` or `ode_parameters` rather than a structural type like
+        `float`.
+    leaf_template
+        A string template to be filled by parameters that represents a single
+        dataset.
+    key_args
+        Key-value pairs to populate the leaf_template and determine nesting
+        of a dataset.
+
+    """
+    root: Path
+    disk_format: str
+    data_type: str
+    leaf_name: str
+    path_name: Optional[str]
+
+
+class MetadataKey(NamedTuple):
+    root: Path
+    disk_format: str
+    data_type: str
+
+
+class DatasetType:
+
+    def __init__(self, name: str,
+                 leaf_template: str, path_template: str = None,
+                 root: Path = None, disk_format: str = None):
+        self.name = name
+        if path_template is not None and path_template not in PATH_TEMPLATES:
+            raise ValueError(f'Invalid path_template specification: {path_template} for DatasetType {self.name}. '
+                             f'path_template must be one of {PATH_TEMPLATES}.')
+        self.path_template = path_template
+        if leaf_template not in LEAF_TEMPLATES:
+            raise ValueError(f'Invalid leaf_template specification: {leaf_template} for DatasetType {self.name}. '
+                             f'leaf_template must be one of {LEAF_TEMPLATES}.')
+        self.leaf_template = leaf_template
+
+        self.root = root
+        # Disk format validated upstream
+        self.disk_format = disk_format
+
+    def __get__(self, instance: 'DataRoot', owner=None) -> 'DatasetType':
+        return type(self)(self.name, self.leaf_template, self.path_template, instance._root, instance._data_format)
+
+    def __call__(self, *args, **key_kwargs) -> DatasetKey:
+        if args:
+            raise ValueError
+
+        leaf_name = self.leaf_template.format(**key_kwargs)
+        path_name = self.path_template.format(**key_kwargs) if self.path_template else None
+        return DatasetKey(self.root, self.disk_format, self.name, leaf_name, path_name)
+
+    def __repr__(self):
+        return f'{type(self).__name__}({", ".join(["=".join([k, str(v)]) for k, v in self.__dict__.items()])})'
+
+
+class MetadataType:
+    def __init__(self, name: str,
```

Comment:
These only have yaml implementations now and are presumed to be at the top level under the date.version root.